File: LeetCode/Medium/0003.LongestSubstringWithoutRepeatingCharacters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> map;
        int ans = 0; int len = s.length();
        
        for(int j = 0, i = 0; j < len; ++j) {
            if(map.find(s[j]) != map.end()) {
                i = max(i, map.at(s[j]));
            }
            
            ans = max(ans, j - i + 1);
            map[s[j]] = j + 1;
        } 
        
        return ans;
    }
};

File: LeetCode/Medium/0003. Longest Substring Without Repeating Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int

        idea:
        have a set to keep track of the chars in current substr
        if a repeated char is found, drop all the chars before the repeated char
        """

        ans = 0
        starting_index = 0
        substr = set()
        for i in range(0, len(s)):
            if s[i] not in substr:
                substr.add(s[i])
            else:
                for j in range(starting_index, i):
                    if s[j] == s[i]: 
                        del_index = j
                
                for j in range(starting_index, del_index):
                    substr.remove(s[j])
                
                starting_index = del_index + 1
            
            ans = max(ans, len(substr))
        
        return ans
